Sources/Core/Shared/GroupOperation.swift

```diff
@@ -225,12 +305,41 @@ public class GroupOperation: Operation, OperationQueueDelegate {
     }
 
     public func operationQueue(queue: OperationQueue, didFinishOperation operation: NSOperation, withErrors errors: [ErrorType]) {
+        guard queue === self.queue else { return }
 
         if operation === finishingOperation {
             finish(fatalErrors)
             queue.suspended = true
         }
     }
+
+    public func operationQueue(queue: OperationQueue, willProduceOperation operation: NSOperation) {
+        guard queue === self.queue else { return }
+
+        // Ensure that produced operations are added to GroupOperation's
+        // internal operations array (and cancelled if appropriate)
+
+        let shouldContinue = groupFinishLock.withCriticalScope { () -> Bool in
+            assert(!finishingOperation.finished, "Cannot produce new operations within a group after the group has completed.")
+            guard !isGroupFinishing else {
+                assertionFailure("Cannot produce new operations within a group after the group has started to finish.")
+                return false
+            }
+            dispatch_group_enter(isAddingOperationsGroup)
+            return true
+        }
+
+        guard shouldContinue else { return }
```

Comment:
I can see this is a common pattern. Returning a bool from a critical scope with a success block, then `guard` on the returned value & exit. As opposed to:

```swift
guard let check = lock.withCriticalScope { () -> Bool in
    guard checkSomethingSafely else { return false }
    doSomethingElse()
    return true
} else { return }
```

I'm just wondering if we can extend `NS(Recursive)Lock` to accept a boolean which we wrap in an `@autoclosure` to do this for us?

Something like:

```swift
extension NSRecursiveLock {
    func guardWithCriticalScope(@autoclosure check: () -> Bool, assertionMessage: String? = .None, success: @noescape (() -> Void)? = .None) -> Bool {
        return withCriticalScope {
            guard check() else {
                if let message = assertionMessage {
                    assertionFailure(message)
                }
                return false
            }
            success?()
            return true
        }
    }
}
```

Usage would then be something like:

```swift
guard let shouldContinue = lock.guardWithCriticalScope(!isGroupFinishing, success: { dispatch_group_enter(isAddingOperationsGroup) }) else { return }
```

Still means you have a trailing closure with you might then want to ` else { }` from... which I'm not 100% sure about.

Just an idea. This same pattern is used here a couple of times and in `Operation`.